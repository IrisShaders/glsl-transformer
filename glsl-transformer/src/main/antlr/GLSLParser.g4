/* GLSL grammar that extends the graphicsfuzz grammar
but is specifically built for this project*/

/*
 * Copyright 2018 The GraphicsFuzz Project Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

parser grammar GLSLParser;

@header {
import io.github.douira.glsl_transformer.parser.ExtendedParser;
}

options {
	tokenVocab = GLSLLexer;
	superClass = ExtendedParser;
}

//the root rule
translationUnit: versionStatement? externalDeclaration* EOF;

//allows for EXT_null_initializer
versionStatement:
	NR NR_VERSION version = (
		NR_GLSL_110
		| NR_GLSL_120
		| NR_GLSLES_100
		| NR_GLSL_130
		| NR_GLSL_140
		| NR_GLSL_150
		| NR_GLSL_330
		| NR_GLSLES_300
		| NR_GLSLES_310
		| NR_GLSLES_320
		| NR_GLSL_400
		| NR_GLSL_410
		| NR_GLSL_420
		| NR_GLSL_430
		| NR_GLSL_440
		| NR_GLSL_450
		| NR_GLSL_460
	) profile = (NR_CORE | NR_COMPATIBILITY | NR_ES)? NR_EOL;

externalDeclaration:
	functionDefinition
	| declaration
	| pragmaStatement
	| extensionStatement
	| customDirectiveStatement
	| includeStatement
	| layoutDefaults
	| emptyDeclaration;

emptyDeclaration: SEMICOLON;

pragmaStatement:
	NR NR_PRAGMA stdGL = NR_STDGL? (
		type = NR_IDENTIFIER
		| type = (NR_PRAGMA_DEBUG | NR_PRAGMA_OPTIMIZE) NR_LPAREN state = (
			NR_ON
			| NR_OFF
		) NR_RPAREN
		| type = NR_PRAGMA_INVARIANT NR_LPAREN state = NR_ALL NR_RPAREN
	) NR_EOL;

extensionStatement:
	NR NR_EXTENSION extensionName = NR_IDENTIFIER (
		NR_COLON extensionBehavior = (
			NR_REQUIRE
			| NR_ENABLE
			| NR_WARN
			| NR_DISABLE
		)
	)? NR_EOL;

customDirectiveStatement: NR NR_CUSTOM content = C_CONTENT? C_EOL;

includeStatement:
	NR NR_INCLUDE (
		NR_STRING_START content = S_CONTENT? S_STRING_END
		| angleStart = NR_STRING_START_ANGLE content = S_CONTENT_ANGLE? S_STRING_END_ANGLE
	) NR_EOL;

layoutDefaults:
	layoutQualifier layoutMode = (UNIFORM | IN | OUT | BUFFER) SEMICOLON;

functionDefinition: functionPrototype compoundStatement;

//Note: diverges from the spec by explicity adding a method call instead of handling it through postfixExpression in functionIdentifier
expression:
	IDENTIFIER # referenceExpression
	| (
		INT16CONSTANT
		| UINT16CONSTANT
		| INT32CONSTANT
		| UINT32CONSTANT
		| INT64CONSTANT
		| UINT64CONSTANT
		| FLOAT16CONSTANT
		| FLOAT32CONSTANT
		| FLOAT64CONSTANT
		| BOOLCONSTANT
	)																													# literalExpression
	| LPAREN value = expression RPAREN												# groupingExpression
	| left = expression LBRACKET right = expression RBRACKET	# arrayAccessExpression
	| operand = expression DOT_LENGTH_METHOD_CALL							# lengthAccessExpression
	//Note: diverges from the spec by not allowing a prefixExpression as an identifier
	//array-type function identfiers are handled by typeSpecifier
	| (IDENTIFIER | typeSpecifier) LPAREN (
		| VOID
		| parameters += expression (COMMA parameters += expression)*
	) RPAREN																				# functionCallExpression
	| operand = expression DOT member = IDENTIFIER	# memberAccessExpression
	| operand = expression op = (INC_OP | DEC_OP)		# postfixExpression
	| <assoc = right> op = (
		INC_OP
		| DEC_OP
		| PLUS_OP
		| MINUS_OP
		| LOGICAL_NOT_OP
		| BITWISE_NEG_OP
	) operand = expression																											# prefixExpression
	| left = expression op = (TIMES_OP | DIV_OP | MOD_OP) right = expression		# multiplicativeExpression
	| left = expression op = (PLUS_OP | MINUS_OP) right = expression						# additiveExpression
	| left = expression op = (LEFT_OP | RIGHT_OP) right = expression						# shiftExpression
	| left = expression op = (LT_OP | GT_OP | LE_OP | GE_OP) right = expression	# relationalExpression
	| left = expression op = (EQ_OP | NE_OP) right = expression									# equalityExpression
	| left = expression op = BITWISE_AND_OP right = expression									# bitwiseAndExpression
	| left = expression op = BITWISE_XOR_OP right = expression									# bitwiseExclusiveOrExpression
	| left = expression op = BITWISE_OR_OP right = expression										# bitwiseInclusiveOrExpression
	| left = expression op = LOGICAL_AND_OP right = expression									# logicalAndExpression
	| left = expression op = LOGICAL_XOR_OP right = expression									# logicalExclusiveOrExpression
	| left = expression op = LOGICAL_OR_OP right = expression										# logicalInclusiveOrExpression
	| <assoc = right> condition = expression QUERY_OP trueAlternative = expression COLON falseAlternative =
		expression # conditionalExpression
	| <assoc = right> left = expression op = (
		ASSIGN_OP
		| MUL_ASSIGN
		| DIV_ASSIGN
		| MOD_ASSIGN
		| ADD_ASSIGN
		| SUB_ASSIGN
		| LEFT_ASSIGN
		| RIGHT_ASSIGN
		| AND_ASSIGN
		| XOR_ASSIGN
		| OR_ASSIGN
	) right = expression													# assignmentExpression
	| left = expression COMMA right = expression	# sequenceExpression;

declaration:
	functionPrototype SEMICOLON # functionDeclaration
	| typeQualifier blockName = IDENTIFIER structBody (
		variableName = IDENTIFIER arraySpecifier?
	)? SEMICOLON # interfaceBlockDeclaration
	| typeQualifier (
		variableNames += IDENTIFIER (COMMA variableNames += IDENTIFIER)*
	)? SEMICOLON																						# variableDeclaration
	| PRECISION precisionQualifier typeSpecifier SEMICOLON	# precisionDeclaration
	| fullySpecifiedType (
		declarationMembers += declarationMember (
			COMMA declarationMembers += declarationMember
		)*
	)? SEMICOLON # typeAndInitDeclaration;

//allows for EXT_subgroup_uniform_control_flow
functionPrototype:
	attribute? fullySpecifiedType IDENTIFIER LPAREN functionParameterList RPAREN attribute?;

functionParameterList: (
		parameters += parameterDeclaration (
			COMMA parameters += parameterDeclaration
		)*
	)?;

parameterDeclaration:
	fullySpecifiedType (parameterName = IDENTIFIER arraySpecifier?)?;

//part of GL_EXT_control_flow_attributes
attribute:
	LBRACKET LBRACKET attributes += singleAttribute (
		COMMA attributes += singleAttribute
	)* RBRACKET RBRACKET;

singleAttribute:
	(prefix = IDENTIFIER COLON COLON)? name = IDENTIFIER (
		LPAREN content = expression RPAREN
	)?;

declarationMember: IDENTIFIER arraySpecifier? (ASSIGN_OP initializer)?;

fullySpecifiedType: typeQualifier? typeSpecifier;

//allows for KHR_memory_scope_semantics
storageQualifier:
	CONST
	| IN
	| OUT
	| INOUT
	| CENTROID
	| PATCH
	| SAMPLE
	| UNIFORM
	| VARYING
	| ATTRIBUTE
	| BUFFER
	| SHARED
	| RESTRICT
	| VOLATILE
	| COHERENT
	| READONLY
	| WRITEONLY
	| SUBROUTINE (
		//TYPE_NAME instead of IDENTIFIER in the spec
		LPAREN typeNames += IDENTIFIER (COMMA typeNames += IDENTIFIER)* RPAREN
	)?
	| DEVICECOHERENT
	| QUEUEFAMILYCOHERENT
	| WORKGROUPCOHERENT
	| SUBGROUPCOHERENT
	| NONPRIVATE;

layoutQualifier:
	LAYOUT LPAREN layoutQualifiers += layoutQualifierId (
		COMMA layoutQualifiers += layoutQualifierId
	)* RPAREN;

//constant expression
layoutQualifierId:
	IDENTIFIER (ASSIGN_OP expression)?	# namedLayoutQualifier
	| SHARED														# sharedLayoutQualifier;

precisionQualifier: HIGHP | MEDIUMP | LOWP;

interpolationQualifier: SMOOTH | FLAT | NOPERSPECTIVE;

invariantQualifier: INVARIANT;

preciseQualifier: PRECISE;

typeQualifier: (
		storageQualifier
		| layoutQualifier
		| precisionQualifier
		| interpolationQualifier
		| invariantQualifier
		| preciseQualifier
	)+;

//TYPE_NAME instead of IDENTIFIER in the spec
typeSpecifier: (
		IDENTIFIER
		| builtinTypeSpecifierFixed
		| builtinTypeSpecifierParseable
		| structSpecifier
	) arraySpecifier?;

//needs duplicated rule parts like this or it becomes mutually left-recursive
//constant expressions
arraySpecifier: arraySpecifierSegment+;
arraySpecifierSegment: (LBRACKET expression? RBRACKET);

builtinTypeSpecifierParseable:
	BOOL
	| BVEC2
	| BVEC3
	| BVEC4
	| FLOAT16
	| F16VEC2
	| F16VEC3
	| F16VEC4
	| F16MAT2X2
	| F16MAT2X3
	| F16MAT2X4
	| F16MAT3X2
	| F16MAT3X3
	| F16MAT3X4
	| F16MAT4X2
	| F16MAT4X3
	| F16MAT4X4
	| FLOAT32
	| F32VEC2
	| F32VEC3
	| F32VEC4
	| F32MAT2X2
	| F32MAT2X3
	| F32MAT2X4
	| F32MAT3X2
	| F32MAT3X3
	| F32MAT3X4
	| F32MAT4X2
	| F32MAT4X3
	| F32MAT4X4
	| FLOAT64
	| F64VEC2
	| F64VEC3
	| F64VEC4
	| F64MAT2X2
	| F64MAT2X3
	| F64MAT2X4
	| F64MAT3X2
	| F64MAT3X3
	| F64MAT3X4
	| F64MAT4X2
	| F64MAT4X3
	| F64MAT4X4
	| INT8
	| I8VEC2
	| I8VEC3
	| I8VEC4
	| UINT8
	| U8VEC2
	| U8VEC3
	| U8VEC4
	| INT16
	| I16VEC2
	| I16VEC3
	| I16VEC4
	| UINT16
	| U16VEC2
	| U16VEC3
	| U16VEC4
	| INT32
	| I32VEC2
	| I32VEC3
	| I32VEC4
	| UINT32
	| U32VEC2
	| U32VEC3
	| U32VEC4
	| INT64
	| I64VEC2
	| I64VEC3
	| I64VEC4
	| UINT64
	| U64VEC2
	| U64VEC3
	| U64VEC4;

builtinTypeSpecifierFixed:
	VOID
	| ATOMIC_UINT
	| SAMPLER2D
	| SAMPLER3D
	| SAMPLERCUBE
	| SAMPLER2DSHADOW
	| SAMPLERCUBESHADOW
	| SAMPLER2DARRAY
	| SAMPLER2DARRAYSHADOW
	| SAMPLERCUBEARRAY
	| SAMPLERCUBEARRAYSHADOW
	| ISAMPLER2D
	| ISAMPLER3D
	| ISAMPLERCUBE
	| ISAMPLER2DARRAY
	| ISAMPLERCUBEARRAY
	| USAMPLER2D
	| USAMPLER3D
	| USAMPLERCUBE
	| USAMPLER2DARRAY
	| USAMPLERCUBEARRAY
	| SAMPLER1D
	| SAMPLER1DSHADOW
	| SAMPLER1DARRAY
	| SAMPLER1DARRAYSHADOW
	| ISAMPLER1D
	| ISAMPLER1DARRAY
	| USAMPLER1D
	| USAMPLER1DARRAY
	| SAMPLER2DRECT
	| SAMPLER2DRECTSHADOW
	| ISAMPLER2DRECT
	| USAMPLER2DRECT
	| SAMPLERBUFFER
	| ISAMPLERBUFFER
	| USAMPLERBUFFER
	| SAMPLER2DMS
	| ISAMPLER2DMS
	| USAMPLER2DMS
	| SAMPLER2DMSARRAY
	| ISAMPLER2DMSARRAY
	| USAMPLER2DMSARRAY
	| IMAGE2D
	| IIMAGE2D
	| UIMAGE2D
	| IMAGE3D
	| IIMAGE3D
	| UIMAGE3D
	| IMAGECUBE
	| IIMAGECUBE
	| UIMAGECUBE
	| IMAGEBUFFER
	| IIMAGEBUFFER
	| UIMAGEBUFFER
	| IMAGE1D
	| IIMAGE1D
	| UIMAGE1D
	| IMAGE1DARRAY
	| IIMAGE1DARRAY
	| UIMAGE1DARRAY
	| IMAGE2DRECT
	| IIMAGE2DRECT
	| UIMAGE2DRECT
	| IMAGE2DARRAY
	| IIMAGE2DARRAY
	| UIMAGE2DARRAY
	| IMAGECUBEARRAY
	| IIMAGECUBEARRAY
	| UIMAGECUBEARRAY
	| IMAGE2DMS
	| IIMAGE2DMS
	| UIMAGE2DMS
	| IMAGE2DMSARRAY
	| IIMAGE2DMSARRAY
	| UIMAGE2DMSARRAY;

structSpecifier: STRUCT IDENTIFIER? structBody;

structBody: LBRACE structMember+ RBRACE;

structMember:
	fullySpecifiedType structDeclarators += structDeclarator (
		COMMA structDeclarators += structDeclarator
	)* SEMICOLON;

structDeclarator: IDENTIFIER arraySpecifier?;

initializer:
	expression
	| LBRACE (
		initializers += initializer (COMMA initializers += initializer)* COMMA?
	)? RBRACE;

statement:
	compoundStatement
	| declarationStatement
	| expressionStatement
	| emptyStatement
	| selectionStatement
	| switchStatement
	| caseLabel
	| forStatement
	| whileStatement
	| doWhileStatement
	| jumpStatement
	| demoteStatement;

compoundStatement: LBRACE statement* RBRACE;

declarationStatement: declaration;

expressionStatement: expression SEMICOLON;

emptyStatement: SEMICOLON;

selectionStatement:
	attribute? IF LPAREN condition = expression RPAREN ifTrue = statement (
		ELSE ifFalse = statement
	)?;

iterationCondition:
	fullySpecifiedType name = IDENTIFIER ASSIGN_OP initializer;

switchStatement:
	attribute? SWITCH LPAREN condition = expression RPAREN compoundStatement;

caseLabel:
	CASE expression COLON	# valuedCaseLabel
	| DEFAULT COLON				# defaultCaseLabel;

whileStatement:
	attribute? WHILE LPAREN (
		condition = expression
		| initCondition = iterationCondition
	) RPAREN loopBody = statement;

doWhileStatement:
	attribute? DO loopBody = statement WHILE LPAREN condition = expression RPAREN SEMICOLON;

forStatement:
	attribute? FOR LPAREN (
		emptyStatement
		| expressionStatement
		| declarationStatement
	) (condition = expression | initCondition = iterationCondition)? SEMICOLON incrementer = expression? RPAREN
		loopBody = statement;

jumpStatement:
	CONTINUE SEMICOLON							# continueStatement
	| BREAK SEMICOLON								# breakStatement
	| RETURN expression? SEMICOLON	# returnStatement
	| DISCARD SEMICOLON							# discardStatement;

demoteStatement: DEMOTE SEMICOLON;
