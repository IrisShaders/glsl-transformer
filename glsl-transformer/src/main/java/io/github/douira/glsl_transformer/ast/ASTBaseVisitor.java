package io.github.douira.glsl_transformer.ast;

public class ASTBaseVisitor<R> implements ASTVisitor<R> {
  @Override
  public R startVisit(ASTNode node) {
    return visit(node);
  }

  @Override
  public R visit(ASTNode node) {
    return node.accept(this);
  }

  @Override
  public R visit(R previousResult, ASTNode node) {
    return aggregateResult(previousResult, visit(node));
  }

  @Override
  public R initialResult() {
    return defaultResult();
  }

  @Override
  public R defaultResult() {
    return null;
  }

  public R aggregateResult(R aggregate, R nextResult) {
    return nextResult;
  }
}
