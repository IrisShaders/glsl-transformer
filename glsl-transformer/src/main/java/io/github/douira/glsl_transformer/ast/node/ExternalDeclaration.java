package io.github.douira.glsl_transformer.ast.node;

import io.github.douira.glsl_transformer.ast.ASTNode;

public abstract class ExternalDeclaration extends ASTNode {
}
