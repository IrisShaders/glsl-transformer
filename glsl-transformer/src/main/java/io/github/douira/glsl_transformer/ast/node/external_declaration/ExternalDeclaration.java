package io.github.douira.glsl_transformer.ast.node.external_declaration;

import io.github.douira.glsl_transformer.ast.node.abstract_node.InnerASTNode;
import io.github.douira.glsl_transformer.ast.query.Root;
import io.github.douira.glsl_transformer.ast.traversal.*;

public abstract class ExternalDeclaration extends InnerASTNode {
  public enum ExternalDeclarationType {
    FUNCTION_DEFINITION,
    DECLARATION,
    PRAGMA_STATEMENT,
    EXTENSION_STATEMENT,
    CUSTOM_DIRECTIVE_STATEMENT,
    INCLUDE_STATEMENT,
    LAYOUT_DEFAULTS,
    EMPTY_DECLARATION
  }

  public abstract ExternalDeclarationType getExternalDeclarationType();

  public abstract <R> R externalDeclarationAccept(ASTVisitor<R> visitor);

  @Override
  public <R> R accept(ASTVisitor<R> visitor) {
    return visitor.aggregateResult(
        visitor.visitExternalDeclaration(this),
        externalDeclarationAccept(visitor));
  }

  @Override
  public void enterNode(ASTListener listener) {
    listener.enterExternalDeclaration(this);
  }

  @Override
  public void exitNode(ASTListener listener) {
    listener.exitExternalDeclaration(this);
  }

  @Override
  public abstract ExternalDeclaration clone();

  @Override
  public ExternalDeclaration cloneInto(Root root) {
    return (ExternalDeclaration) super.cloneInto(root);
  }

  @Override
  public ExternalDeclaration cloneSeparate() {
    return (ExternalDeclaration) super.cloneSeparate();
  }
}
