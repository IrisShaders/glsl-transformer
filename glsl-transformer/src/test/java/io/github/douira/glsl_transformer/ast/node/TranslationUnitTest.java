package io.github.douira.glsl_transformer.ast.node;

import static org.junit.jupiter.api.Assertions.*;

import java.util.stream.*;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;

import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.annotations.SnapshotName;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import io.github.douira.glsl_transformer.GLSLParser;
import io.github.douira.glsl_transformer.ast.node.external_declaration.ExternalDeclaration;
import io.github.douira.glsl_transformer.ast.transform.*;
import io.github.douira.glsl_transformer.test_util.*;
import io.github.douira.glsl_transformer.test_util.TestCaseProvider.Spacing;

@ExtendWith({ SnapshotExtension.class })
public class TranslationUnitTest extends TestWithASTTransformer {
  private Expect expect;

  @Test
  void testInjection() {
    transformer.setTransformation(tree -> tree.injectNode(
        ASTInjectionPoint.BEFORE_FUNCTIONS,
        transformer.parseNode(
            "float x;",
            tree,
            GLSLParser::externalDeclaration,
            ASTBuilder::visitExternalDeclaration)));
    assertEquals(
        "int a;\nfloat x;\nvoid main() {\n}\n",
        transformer.transform("int a;\nvoid main() {\n}\n"));
    assertEquals(
        "int a;\nfloat x;\n",
        transformer.transform("int a;\n"));
  }

  @ParameterizedTest
  @TestCaseSource(caseSet = "testInjectExDecl", spacing = Spacing.TRIM_SINGLE_BOTH)
  @SnapshotName("testInject")
  void testInject(String scenario, String input) {
    for (var location : ASTInjectionPoint.values()) {
      transformer.setTransformation(tree -> tree.injectNode(
          location,
          Stream.of("x", "y")
              .<ExternalDeclaration>map(name -> transformer.parseNode(
                  "float " + name + ";",
                  tree,
                  GLSLParser::externalDeclaration,
                  ASTBuilder::visitExternalDeclaration))
              .collect(Collectors.toList())));
      expect
          .scenario(scenario + "/" + location.toString().toLowerCase())
          .toMatchSnapshot(SnapshotUtil.inputOutputSnapshot(
              input, transformer.transform(input)));
    }
  }
}
