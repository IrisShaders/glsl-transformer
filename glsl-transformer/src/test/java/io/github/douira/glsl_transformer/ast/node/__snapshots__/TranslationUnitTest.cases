§testInjectExDecl empty:

§testInjectExDecl single_version:
#version 5432

§testInjectExDecl single_extension:
#extension extension_name: require

§testInjectExDecl single_declaration:
varying vec2 a;

§testInjectExDecl single_layout:
layout(location = 0) out vec3 layoutOne;

§testInjectExDecl single_pragma:
#pragma foo

§testInjectExDecl single_function:
void main() {}

§testInjectExDecl full:
#version 5432
#extension extension_name: require
#pragma foo
varying vec2 a;
void functionDef(const in float param) {
  functionCall();
}

§testInjectExDecl full_reverse:
#version 5432
void functionDef(const in float param) {
  functionCall();
}
varying vec2 a;
#pragma foo
#extension extension_name: require

§testInjectExDecl version_with_pragma:
#version 5432
#pragma foo

§testInjectExDecl version_with_extension:
#version 5432
#extension extension_name: require

§testInjectExDecl unparsed_separators:
#version 5432
varying vec2 a;

§testInjectExDecl unparsed_and_functions:
varying vec2 a;
varying vec2 b;
void funcA() {
}
void funcB() {
}
void funcC() {
}
