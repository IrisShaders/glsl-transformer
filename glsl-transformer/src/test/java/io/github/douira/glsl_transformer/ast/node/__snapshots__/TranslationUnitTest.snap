testInject[empty/before_all]=[

<><><><><><><><><><><><><><><><><><><><><><><><><>
float x;
float y;

]


testInject[empty/before_declarations]=[

<><><><><><><><><><><><><><><><><><><><><><><><><>
float x;
float y;

]


testInject[empty/before_functions]=[

<><><><><><><><><><><><><><><><><><><><><><><><><>
float x;
float y;

]


testInject[empty/end]=[

<><><><><><><><><><><><><><><><><><><><><><><><><>
float x;
float y;

]


testInject[full/before_all]=[
#version 5432
#extension extension_name: require
#pragma foo
varying vec2 a;
void functionDef(const in float param) {
  functionCall();
}

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
float x;
float y;
#extension extension_name: require
#pragma foo
varying vec2 a;
void functionDef(const in float param) {
	functionCall();
}

]


testInject[full/before_declarations]=[
#version 5432
#extension extension_name: require
#pragma foo
varying vec2 a;
void functionDef(const in float param) {
  functionCall();
}

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
#extension extension_name: require
#pragma foo
float x;
float y;
varying vec2 a;
void functionDef(const in float param) {
	functionCall();
}

]


testInject[full/before_functions]=[
#version 5432
#extension extension_name: require
#pragma foo
varying vec2 a;
void functionDef(const in float param) {
  functionCall();
}

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
#extension extension_name: require
#pragma foo
varying vec2 a;
float x;
float y;
void functionDef(const in float param) {
	functionCall();
}

]


testInject[full/end]=[
#version 5432
#extension extension_name: require
#pragma foo
varying vec2 a;
void functionDef(const in float param) {
  functionCall();
}

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
#extension extension_name: require
#pragma foo
varying vec2 a;
void functionDef(const in float param) {
	functionCall();
}
float x;
float y;

]


testInject[full_reverse/before_all]=[
#version 5432
void functionDef(const in float param) {
  functionCall();
}
varying vec2 a;
#pragma foo
#extension extension_name: require

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
float x;
float y;
void functionDef(const in float param) {
	functionCall();
}
varying vec2 a;
#pragma foo
#extension extension_name: require

]


testInject[full_reverse/before_declarations]=[
#version 5432
void functionDef(const in float param) {
  functionCall();
}
varying vec2 a;
#pragma foo
#extension extension_name: require

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
float x;
float y;
void functionDef(const in float param) {
	functionCall();
}
varying vec2 a;
#pragma foo
#extension extension_name: require

]


testInject[full_reverse/before_functions]=[
#version 5432
void functionDef(const in float param) {
  functionCall();
}
varying vec2 a;
#pragma foo
#extension extension_name: require

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
float x;
float y;
void functionDef(const in float param) {
	functionCall();
}
varying vec2 a;
#pragma foo
#extension extension_name: require

]


testInject[full_reverse/end]=[
#version 5432
void functionDef(const in float param) {
  functionCall();
}
varying vec2 a;
#pragma foo
#extension extension_name: require

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
void functionDef(const in float param) {
	functionCall();
}
varying vec2 a;
#pragma foo
#extension extension_name: require
float x;
float y;

]


testInject[single_declaration/before_all]=[
varying vec2 a;

<><><><><><><><><><><><><><><><><><><><><><><><><>
float x;
float y;
varying vec2 a;

]


testInject[single_declaration/before_declarations]=[
varying vec2 a;

<><><><><><><><><><><><><><><><><><><><><><><><><>
float x;
float y;
varying vec2 a;

]


testInject[single_declaration/before_functions]=[
varying vec2 a;

<><><><><><><><><><><><><><><><><><><><><><><><><>
varying vec2 a;
float x;
float y;

]


testInject[single_declaration/end]=[
varying vec2 a;

<><><><><><><><><><><><><><><><><><><><><><><><><>
varying vec2 a;
float x;
float y;

]


testInject[single_extension/before_all]=[
#extension extension_name: require

<><><><><><><><><><><><><><><><><><><><><><><><><>
float x;
float y;
#extension extension_name: require

]


testInject[single_extension/before_declarations]=[
#extension extension_name: require

<><><><><><><><><><><><><><><><><><><><><><><><><>
#extension extension_name: require
float x;
float y;

]


testInject[single_extension/before_functions]=[
#extension extension_name: require

<><><><><><><><><><><><><><><><><><><><><><><><><>
#extension extension_name: require
float x;
float y;

]


testInject[single_extension/end]=[
#extension extension_name: require

<><><><><><><><><><><><><><><><><><><><><><><><><>
#extension extension_name: require
float x;
float y;

]


testInject[single_function/before_all]=[
void main() {}

<><><><><><><><><><><><><><><><><><><><><><><><><>
float x;
float y;
void main() {
}

]


testInject[single_function/before_declarations]=[
void main() {}

<><><><><><><><><><><><><><><><><><><><><><><><><>
float x;
float y;
void main() {
}

]


testInject[single_function/before_functions]=[
void main() {}

<><><><><><><><><><><><><><><><><><><><><><><><><>
float x;
float y;
void main() {
}

]


testInject[single_function/end]=[
void main() {}

<><><><><><><><><><><><><><><><><><><><><><><><><>
void main() {
}
float x;
float y;

]


testInject[single_layout/before_all]=[
layout(location = 0) out vec3 layoutOne;

<><><><><><><><><><><><><><><><><><><><><><><><><>
float x;
float y;
layout(location = 0) out vec3 layoutOne;

]


testInject[single_layout/before_declarations]=[
layout(location = 0) out vec3 layoutOne;

<><><><><><><><><><><><><><><><><><><><><><><><><>
float x;
float y;
layout(location = 0) out vec3 layoutOne;

]


testInject[single_layout/before_functions]=[
layout(location = 0) out vec3 layoutOne;

<><><><><><><><><><><><><><><><><><><><><><><><><>
layout(location = 0) out vec3 layoutOne;
float x;
float y;

]


testInject[single_layout/end]=[
layout(location = 0) out vec3 layoutOne;

<><><><><><><><><><><><><><><><><><><><><><><><><>
layout(location = 0) out vec3 layoutOne;
float x;
float y;

]


testInject[single_pragma/before_all]=[
#pragma foo

<><><><><><><><><><><><><><><><><><><><><><><><><>
float x;
float y;
#pragma foo

]


testInject[single_pragma/before_declarations]=[
#pragma foo

<><><><><><><><><><><><><><><><><><><><><><><><><>
#pragma foo
float x;
float y;

]


testInject[single_pragma/before_functions]=[
#pragma foo

<><><><><><><><><><><><><><><><><><><><><><><><><>
#pragma foo
float x;
float y;

]


testInject[single_pragma/end]=[
#pragma foo

<><><><><><><><><><><><><><><><><><><><><><><><><>
#pragma foo
float x;
float y;

]


testInject[single_version/before_all]=[
#version 5432

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
float x;
float y;

]


testInject[single_version/before_declarations]=[
#version 5432

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
float x;
float y;

]


testInject[single_version/before_functions]=[
#version 5432

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
float x;
float y;

]


testInject[single_version/end]=[
#version 5432

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
float x;
float y;

]


testInject[unparsed_and_functions/before_all]=[
varying vec2 a;
varying vec2 b;
void funcA() {
}
void funcB() {
}
void funcC() {
}
<><><><><><><><><><><><><><><><><><><><><><><><><>
float x;
float y;
varying vec2 a;
varying vec2 b;
void funcA() {
}
void funcB() {
}
void funcC() {
}

]


testInject[unparsed_and_functions/before_declarations]=[
varying vec2 a;
varying vec2 b;
void funcA() {
}
void funcB() {
}
void funcC() {
}
<><><><><><><><><><><><><><><><><><><><><><><><><>
float x;
float y;
varying vec2 a;
varying vec2 b;
void funcA() {
}
void funcB() {
}
void funcC() {
}

]


testInject[unparsed_and_functions/before_functions]=[
varying vec2 a;
varying vec2 b;
void funcA() {
}
void funcB() {
}
void funcC() {
}
<><><><><><><><><><><><><><><><><><><><><><><><><>
varying vec2 a;
varying vec2 b;
float x;
float y;
void funcA() {
}
void funcB() {
}
void funcC() {
}

]


testInject[unparsed_and_functions/end]=[
varying vec2 a;
varying vec2 b;
void funcA() {
}
void funcB() {
}
void funcC() {
}
<><><><><><><><><><><><><><><><><><><><><><><><><>
varying vec2 a;
varying vec2 b;
void funcA() {
}
void funcB() {
}
void funcC() {
}
float x;
float y;

]


testInject[unparsed_separators/before_all]=[
#version 5432
varying vec2 a;

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
float x;
float y;
varying vec2 a;

]


testInject[unparsed_separators/before_declarations]=[
#version 5432
varying vec2 a;

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
float x;
float y;
varying vec2 a;

]


testInject[unparsed_separators/before_functions]=[
#version 5432
varying vec2 a;

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
varying vec2 a;
float x;
float y;

]


testInject[unparsed_separators/end]=[
#version 5432
varying vec2 a;

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
varying vec2 a;
float x;
float y;

]


testInject[version_with_extension/before_all]=[
#version 5432
#extension extension_name: require

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
float x;
float y;
#extension extension_name: require

]


testInject[version_with_extension/before_declarations]=[
#version 5432
#extension extension_name: require

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
#extension extension_name: require
float x;
float y;

]


testInject[version_with_extension/before_functions]=[
#version 5432
#extension extension_name: require

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
#extension extension_name: require
float x;
float y;

]


testInject[version_with_extension/end]=[
#version 5432
#extension extension_name: require

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
#extension extension_name: require
float x;
float y;

]


testInject[version_with_pragma/before_all]=[
#version 5432
#pragma foo

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
float x;
float y;
#pragma foo

]


testInject[version_with_pragma/before_declarations]=[
#version 5432
#pragma foo

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
#pragma foo
float x;
float y;

]


testInject[version_with_pragma/before_functions]=[
#version 5432
#pragma foo

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
#pragma foo
float x;
float y;

]


testInject[version_with_pragma/end]=[
#version 5432
#pragma foo

<><><><><><><><><><><><><><><><><><><><><><><><><>
#version 5432
#pragma foo
float x;
float y;

]